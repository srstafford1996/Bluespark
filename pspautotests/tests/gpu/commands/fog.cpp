#include <common.h>
#include <pspdisplay.h>
#include <pspgu.h>
#include <pspkernel.h>

extern "C" int sceDmacMemcpy(void *dest, const void *source, unsigned int size);

#define GU_IMM_FOG 0x00400000

typedef struct {
	u32 color;
	float x, y, z;
} VertexColorF32;

static u8 *fbp0 = 0;
static u8 *dbp0 = fbp0 + 512 * 272 * sizeof(u32);

static u32 copybuf[32 * 1];
static unsigned int __attribute__((aligned(16))) list[131072];
static __attribute__((aligned(16))) VertexColorF32 vertices_f32[256];

inline VertexColorF32 makeVertex32(u32 c, float x, float y, float z) {
	VertexColorF32 v;
	v.color = c;
	v.x = x;
	v.y = y;
	v.z = z;
	return v;
}

void displayBuffer(const char *reason) {
	sceKernelDcacheWritebackInvalidateRange(copybuf, sizeof(copybuf));
	sceDmacMemcpy(copybuf, sceGeEdramGetAddr(), sizeof(copybuf));
	const u32 *buf = copybuf;

	checkpoint("%s: COLOR=%08x", reason, buf[0]);

	// Reset.
	memset(copybuf, 0x44, sizeof(copybuf));
	sceKernelDcacheWritebackInvalidateRange(copybuf, sizeof(copybuf));
	sceDmacMemcpy(sceGeEdramGetAddr(), copybuf, sizeof(copybuf));
}

void drawBoxCommands(u32 c) {
	vertices_f32[0] = makeVertex32(c, -1.0, -1.0, 0.0);
	vertices_f32[1] = makeVertex32(c, 1.0, 1.0, 0.0);
	sceKernelDcacheWritebackInvalidateRange(vertices_f32, sizeof(vertices_f32));

	sceGuDrawArray(GU_SPRITES, GU_COLOR_8888 | GU_VERTEX_32BITF | GU_TRANSFORM_3D, 2, NULL, vertices_f32);
}

void drawBoxCommandsImm(u32 c, u8 fogfactor) {
	sceGuSendCommandi(18, GU_COLOR_8888 | GU_VERTEX_32BITF | GU_TRANSFORM_3D);

	sceGuSendCommandi(0xF0, (2048 - (480 / 2)) << 4);
	sceGuSendCommandi(0xF1, (2048 - (272 / 2)) << 4);
	sceGuSendCommandi(0xF2, 0);
	sceGuSendCommandi(0xF6, c & 0x00FFFFFF);
	sceGuSendCommandi(0xF9, 0);
	sceGuSendCommandi(0xF8, fogfactor);
	sceGuSendCommandi(0xF7, GU_IMM_FOG | (GU_SPRITES << 8) | (c >> 24));

	sceGuSendCommandi(0xF0, (2048 + (480 / 2)) << 4);
	sceGuSendCommandi(0xF1, (2048 + (272 / 2)) << 4);
	sceGuSendCommandi(0xF7, GU_IMM_FOG | (7 << 8) | (c >> 24));
}

void init() { 
	sceGuInit();
	sceGuStart(GU_DIRECT, list);
	sceGuDrawBuffer(GU_PSM_8888, fbp0, 512);
	sceGuDispBuffer(480, 272, fbp0, 512);
	sceGuDepthBuffer(dbp0, 512);
	sceGuOffset(2048 - (480 / 2), 2048 - (272 / 2));
	sceGuViewport(2048, 2048, 480, 272);
	sceGuDepthRange(65535, 0);
	sceGuDepthMask(0);
	sceGuScissor(0, 0, 10, 10);
	sceGuEnable(GU_SCISSOR_TEST);
	sceGuFrontFace(GU_CW);
	sceGuShadeModel(GU_SMOOTH);
	sceGuDisable(GU_TEXTURE_2D);

	ScePspFMatrix4 ones = {
		{1, 0, 0, 0},
		{0, 1, 0, 0},
		{0, 0, 1, 0},
		{0, 0, 0, 1},
	};

	sceGuSetMatrix(GU_MODEL, &ones);
	sceGuSetMatrix(GU_VIEW, &ones);
	sceGuSetMatrix(GU_PROJECTION, &ones);

	sceGuFinish();
	sceGuSync(0, 0);
 
	sceDisplayWaitVblankStart();
	sceGuDisplay(1);

	memset(copybuf, 0x44, sizeof(copybuf));
	sceKernelDcacheWritebackInvalidateRange(copybuf, sizeof(copybuf));
	sceDmacMemcpy(sceGeEdramGetAddr(), copybuf, sizeof(copybuf));

	displayBuffer("Initial");
}

void testFog(const char *title, u32 prev, u32 c, u32 fogc, float fog_near, float fog_far) {
	for (size_t i = 0; i < sizeof(copybuf) / 4; ++i) {
		copybuf[i] = prev;
	}
	sceKernelDcacheWritebackInvalidateRange(copybuf, sizeof(copybuf));
	sceDmacMemcpy(sceGeEdramGetAddr(), copybuf, sizeof(copybuf));

	sceGuStart(GU_DIRECT, list);

	sceGuDisable(GU_BLEND);
	sceGuDisable(GU_STENCIL_TEST);
	sceGuEnable(GU_FOG);

	sceGuFog(fog_near, fog_far, fogc);

	drawBoxCommands(c);

	sceGuFinish();
	sceGuSync(GU_SYNC_WAIT, GU_SYNC_WHAT_DONE);

	displayBuffer(title);
}

void testImmFog(const char *title, u32 prev, u32 c, u32 fogc, u8 fogfactor) {
	for (size_t i = 0; i < sizeof(copybuf) / 4; ++i) {
		copybuf[i] = prev;
	}
	sceKernelDcacheWritebackInvalidateRange(copybuf, sizeof(copybuf));
	sceDmacMemcpy(sceGeEdramGetAddr(), copybuf, sizeof(copybuf));

	sceGuStart(GU_DIRECT, list);

	sceGuDisable(GU_BLEND);
	sceGuDisable(GU_STENCIL_TEST);
	sceGuEnable(GU_FOG);

	sceGuFog(0.0f, 1.0f, fogc);

	drawBoxCommandsImm(c, fogfactor);

	sceGuFinish();
	sceGuSync(GU_SYNC_WAIT, GU_SYNC_WHAT_DONE);

	displayBuffer(title);
}

extern volatile int didResched;

extern "C" int main(int argc, char *argv[]) {
	init();

	schedf("framebuf: %08x\n", sceDisplaySetFrameBuf(sceGeEdramGetAddr(), 512, GU_PSM_8888, PSP_DISPLAY_SETBUF_IMMEDIATE));
	schedf("dispmode: %08x\n", sceDisplaySetMode(0, 480, 272));

	checkpointNext("Common:");
	testFog("  Basic", 0x44444444, 0x88888888, 0xFFFFFFFF, 1.0f, 1.0f);
	testFog("  Near neg", 0x44444444, 0x88888888, 0xFFFFFFFF, -1.0f, 1.0f);
	testFog("  Far neg", 0x44444444, 0x88888888, 0xFFFFFFFF, 1.0f, -1.0f);
	testFog("  Both neg", 0x44444444, 0x88888888, 0xFFFFFFFF, -1.0f, -1.0f);
	testFog("  Near zero", 0x44444444, 0x88888888, 0xFFFFFFFF, 0.0f, 1.0f);
	testFog("  Far zero", 0x44444444, 0x88888888, 0xFFFFFFFF, 1.0f, 0.0f);
	testImmFog("  Direct fog 00", 0x44444444, 0x88888888, 0xFFFFFFFF, 0x00);
	testImmFog("  Direct fog 80", 0x44444444, 0x88888888, 0xFFFFFFFF, 0x80);
	testImmFog("  Direct fog FF", 0x44444444, 0x88888888, 0xFFFFFFFF, 0xFF);

	checkpointNext("Rounding:");
	for (int i = 0; i < 256; ++i) {
		char temp[128];
		sprintf(temp, "  Fog value %02x", i);
		testImmFog(temp, 0x44444444, 0x11881100, 0xFFFF33FF, i);
	}
	// More exhaustive test.
	/*for (int i = 0; i < 256; ++i) {
		char temp[128];
		snprintf(temp, 128, "  0x7F * 0x%02X - Zero (%f / %d)", i, i * 127.0 / 255.0, (i * 127) % 255);

		for (int j = 0; j < 256; ++j) {
			int jjj = j | (j << 8) | (j << 16);
			snprintf(temp, 128, "%d,%d", i, j);
			testImmFog(temp, 0x00000000, jjj, 0x00FFFFFF, (u8)i);
		}
		if (i != 255) {
			flushschedf();
			didResched = 0;
		}
	}*/

	sceGuTerm();

	return 0;
}
