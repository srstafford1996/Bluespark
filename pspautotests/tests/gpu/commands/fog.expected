[x] Initial: COLOR=44444444
framebuf: 00000000
dispmode: 00000000

[x] Common:
[x]   Basic: COLOR=44ffffff
[x]   Near neg: COLOR=44c3c3c3
[x]   Far neg: COLOR=44c3c3c3
[x]   Both neg: COLOR=44ffffff
[x]   Near zero: COLOR=44888888
[x]   Far zero: COLOR=44ffffff
[x]   Direct fog 00: COLOR=44ffffff
[x]   Direct fog 80: COLOR=44c3c3c3
[x]   Direct fog FF: COLOR=44888888

[x] Rounding:
[x]   Fog value 00: COLOR=44ff33ff
[x]   Fog value 01: COLOR=44fe33fe
[x]   Fog value 02: COLOR=44fe33fd
[x]   Fog value 03: COLOR=44fd33fc
[x]   Fog value 04: COLOR=44fd33fb
[x]   Fog value 05: COLOR=44fc33fa
[x]   Fog value 06: COLOR=44fc33f9
[x]   Fog value 07: COLOR=44fb32f8
[x]   Fog value 08: COLOR=44fb32f7
[x]   Fog value 09: COLOR=44fa32f6
[x]   Fog value 0a: COLOR=44fa32f5
[x]   Fog value 0b: COLOR=44f932f4
[x]   Fog value 0c: COLOR=44f932f3
[x]   Fog value 0d: COLOR=44f832f2
[x]   Fog value 0e: COLOR=44f831f1
[x]   Fog value 0f: COLOR=44f831f0
[x]   Fog value 10: COLOR=44f731ef
[x]   Fog value 11: COLOR=44f731ee
[x]   Fog value 12: COLOR=44f631ed
[x]   Fog value 13: COLOR=44f631ec
[x]   Fog value 14: COLOR=44f531eb
[x]   Fog value 15: COLOR=44f531ea
[x]   Fog value 16: COLOR=44f430e9
[x]   Fog value 17: COLOR=44f430e8
[x]   Fog value 18: COLOR=44f330e7
[x]   Fog value 19: COLOR=44f330e6
[x]   Fog value 1a: COLOR=44f230e5
[x]   Fog value 1b: COLOR=44f230e4
[x]   Fog value 1c: COLOR=44f130e3
[x]   Fog value 1d: COLOR=44f12fe2
[x]   Fog value 1e: COLOR=44f12fe1
[x]   Fog value 1f: COLOR=44f02fe0
[x]   Fog value 20: COLOR=44f02fdf
[x]   Fog value 21: COLOR=44ef2fde
[x]   Fog value 22: COLOR=44ef2fdd
[x]   Fog value 23: COLOR=44ee2fdc
[x]   Fog value 24: COLOR=44ee2fdb
[x]   Fog value 25: COLOR=44ed2eda
[x]   Fog value 26: COLOR=44ed2ed9
[x]   Fog value 27: COLOR=44ec2ed8
[x]   Fog value 28: COLOR=44ec2ed7
[x]   Fog value 29: COLOR=44eb2ed6
[x]   Fog value 2a: COLOR=44eb2ed5
[x]   Fog value 2b: COLOR=44eb2ed4
[x]   Fog value 2c: COLOR=44ea2dd3
[x]   Fog value 2d: COLOR=44ea2dd2
[x]   Fog value 2e: COLOR=44e92dd1
[x]   Fog value 2f: COLOR=44e92dd0
[x]   Fog value 30: COLOR=44e82dcf
[x]   Fog value 31: COLOR=44e82dce
[x]   Fog value 32: COLOR=44e72dcd
[x]   Fog value 33: COLOR=44e72dcc
[x]   Fog value 34: COLOR=44e62ccb
[x]   Fog value 35: COLOR=44e62cca
[x]   Fog value 36: COLOR=44e52cc9
[x]   Fog value 37: COLOR=44e52cc8
[x]   Fog value 38: COLOR=44e42cc7
[x]   Fog value 39: COLOR=44e42cc6
[x]   Fog value 3a: COLOR=44e42cc5
[x]   Fog value 3b: COLOR=44e32bc4
[x]   Fog value 3c: COLOR=44e32bc3
[x]   Fog value 3d: COLOR=44e22bc2
[x]   Fog value 3e: COLOR=44e22bc1
[x]   Fog value 3f: COLOR=44e12bc0
[x]   Fog value 40: COLOR=44e12bbf
[x]   Fog value 41: COLOR=44e02bbe
[x]   Fog value 42: COLOR=44e02bbd
[x]   Fog value 43: COLOR=44df2abc
[x]   Fog value 44: COLOR=44df2abb
[x]   Fog value 45: COLOR=44de2aba
[x]   Fog value 46: COLOR=44de2ab9
[x]   Fog value 47: COLOR=44dd2ab8
[x]   Fog value 48: COLOR=44dd2ab7
[x]   Fog value 49: COLOR=44dd2ab6
[x]   Fog value 4a: COLOR=44dc29b5
[x]   Fog value 4b: COLOR=44dc29b4
[x]   Fog value 4c: COLOR=44db29b3
[x]   Fog value 4d: COLOR=44db29b2
[x]   Fog value 4e: COLOR=44da29b1
[x]   Fog value 4f: COLOR=44da29b0
[x]   Fog value 50: COLOR=44d929af
[x]   Fog value 51: COLOR=44d929ae
[x]   Fog value 52: COLOR=44d828ad
[x]   Fog value 53: COLOR=44d828ac
[x]   Fog value 54: COLOR=44d728ab
[x]   Fog value 55: COLOR=44d728aa
[x]   Fog value 56: COLOR=44d728a9
[x]   Fog value 57: COLOR=44d628a8
[x]   Fog value 58: COLOR=44d628a7
[x]   Fog value 59: COLOR=44d527a6
[x]   Fog value 5a: COLOR=44d527a5
[x]   Fog value 5b: COLOR=44d427a4
[x]   Fog value 5c: COLOR=44d427a3
[x]   Fog value 5d: COLOR=44d327a2
[x]   Fog value 5e: COLOR=44d327a1
[x]   Fog value 5f: COLOR=44d227a0
[x]   Fog value 60: COLOR=44d2279f
[x]   Fog value 61: COLOR=44d1269e
[x]   Fog value 62: COLOR=44d1269d
[x]   Fog value 63: COLOR=44d0269c
[x]   Fog value 64: COLOR=44d0269b
[x]   Fog value 65: COLOR=44d0269a
[x]   Fog value 66: COLOR=44cf2699
[x]   Fog value 67: COLOR=44cf2698
[x]   Fog value 68: COLOR=44ce2597
[x]   Fog value 69: COLOR=44ce2596
[x]   Fog value 6a: COLOR=44cd2595
[x]   Fog value 6b: COLOR=44cd2594
[x]   Fog value 6c: COLOR=44cc2593
[x]   Fog value 6d: COLOR=44cc2592
[x]   Fog value 6e: COLOR=44cb2591
[x]   Fog value 6f: COLOR=44cb2590
[x]   Fog value 70: COLOR=44ca248f
[x]   Fog value 71: COLOR=44ca248e
[x]   Fog value 72: COLOR=44ca248d
[x]   Fog value 73: COLOR=44c9248c
[x]   Fog value 74: COLOR=44c9248b
[x]   Fog value 75: COLOR=44c8248a
[x]   Fog value 76: COLOR=44c82489
[x]   Fog value 77: COLOR=44c72388
[x]   Fog value 78: COLOR=44c72387
[x]   Fog value 79: COLOR=44c62386
[x]   Fog value 7a: COLOR=44c62385
[x]   Fog value 7b: COLOR=44c52384
[x]   Fog value 7c: COLOR=44c52383
[x]   Fog value 7d: COLOR=44c42382
[x]   Fog value 7e: COLOR=44c42381
[x]   Fog value 7f: COLOR=44c32280
[x]   Fog value 80: COLOR=44c3227f
[x]   Fog value 81: COLOR=44c3227e
[x]   Fog value 82: COLOR=44c2227d
[x]   Fog value 83: COLOR=44c2227c
[x]   Fog value 84: COLOR=44c1227b
[x]   Fog value 85: COLOR=44c1227a
[x]   Fog value 86: COLOR=44c02279
[x]   Fog value 87: COLOR=44c02178
[x]   Fog value 88: COLOR=44bf2177
[x]   Fog value 89: COLOR=44bf2176
[x]   Fog value 8a: COLOR=44be2175
[x]   Fog value 8b: COLOR=44be2174
[x]   Fog value 8c: COLOR=44bd2173
[x]   Fog value 8d: COLOR=44bd2172
[x]   Fog value 8e: COLOR=44bc2071
[x]   Fog value 8f: COLOR=44bc2070
[x]   Fog value 90: COLOR=44bc206f
[x]   Fog value 91: COLOR=44bb206e
[x]   Fog value 92: COLOR=44bb206d
[x]   Fog value 93: COLOR=44ba206c
[x]   Fog value 94: COLOR=44ba206b
[x]   Fog value 95: COLOR=44b9206a
[x]   Fog value 96: COLOR=44b91f69
[x]   Fog value 97: COLOR=44b81f68
[x]   Fog value 98: COLOR=44b81f67
[x]   Fog value 99: COLOR=44b71f66
[x]   Fog value 9a: COLOR=44b71f65
[x]   Fog value 9b: COLOR=44b61f64
[x]   Fog value 9c: COLOR=44b61f63
[x]   Fog value 9d: COLOR=44b61e62
[x]   Fog value 9e: COLOR=44b51e61
[x]   Fog value 9f: COLOR=44b51e60
[x]   Fog value a0: COLOR=44b41e5f
[x]   Fog value a1: COLOR=44b41e5e
[x]   Fog value a2: COLOR=44b31e5d
[x]   Fog value a3: COLOR=44b31e5c
[x]   Fog value a4: COLOR=44b21e5b
[x]   Fog value a5: COLOR=44b21d5a
[x]   Fog value a6: COLOR=44b11d59
[x]   Fog value a7: COLOR=44b11d58
[x]   Fog value a8: COLOR=44b01d57
[x]   Fog value a9: COLOR=44b01d56
[x]   Fog value aa: COLOR=44af1d55
[x]   Fog value ab: COLOR=44af1d54
[x]   Fog value ac: COLOR=44af1c53
[x]   Fog value ad: COLOR=44ae1c52
[x]   Fog value ae: COLOR=44ae1c51
[x]   Fog value af: COLOR=44ad1c50
[x]   Fog value b0: COLOR=44ad1c4f
[x]   Fog value b1: COLOR=44ac1c4e
[x]   Fog value b2: COLOR=44ac1c4d
[x]   Fog value b3: COLOR=44ab1c4c
[x]   Fog value b4: COLOR=44ab1b4b
[x]   Fog value b5: COLOR=44aa1b4a
[x]   Fog value b6: COLOR=44aa1b49
[x]   Fog value b7: COLOR=44a91b48
[x]   Fog value b8: COLOR=44a91b47
[x]   Fog value b9: COLOR=44a91b46
[x]   Fog value ba: COLOR=44a81b45
[x]   Fog value bb: COLOR=44a81a44
[x]   Fog value bc: COLOR=44a71a43
[x]   Fog value bd: COLOR=44a71a42
[x]   Fog value be: COLOR=44a61a41
[x]   Fog value bf: COLOR=44a61a40
[x]   Fog value c0: COLOR=44a51a3f
[x]   Fog value c1: COLOR=44a51a3e
[x]   Fog value c2: COLOR=44a41a3d
[x]   Fog value c3: COLOR=44a4193c
[x]   Fog value c4: COLOR=44a3193b
[x]   Fog value c5: COLOR=44a3193a
[x]   Fog value c6: COLOR=44a21939
[x]   Fog value c7: COLOR=44a21938
[x]   Fog value c8: COLOR=44a21937
[x]   Fog value c9: COLOR=44a11936
[x]   Fog value ca: COLOR=44a11835
[x]   Fog value cb: COLOR=44a01834
[x]   Fog value cc: COLOR=44a01833
[x]   Fog value cd: COLOR=449f1832
[x]   Fog value ce: COLOR=449f1831
[x]   Fog value cf: COLOR=449e1830
[x]   Fog value d0: COLOR=449e182f
[x]   Fog value d1: COLOR=449d182e
[x]   Fog value d2: COLOR=449d172d
[x]   Fog value d3: COLOR=449c172c
[x]   Fog value d4: COLOR=449c172b
[x]   Fog value d5: COLOR=449b172a
[x]   Fog value d6: COLOR=449b1729
[x]   Fog value d7: COLOR=449b1728
[x]   Fog value d8: COLOR=449a1727
[x]   Fog value d9: COLOR=449a1626
[x]   Fog value da: COLOR=44991625
[x]   Fog value db: COLOR=44991624
[x]   Fog value dc: COLOR=44981623
[x]   Fog value dd: COLOR=44981622
[x]   Fog value de: COLOR=44971621
[x]   Fog value df: COLOR=44971620
[x]   Fog value e0: COLOR=4496161f
[x]   Fog value e1: COLOR=4496151e
[x]   Fog value e2: COLOR=4495151d
[x]   Fog value e3: COLOR=4495151c
[x]   Fog value e4: COLOR=4495151b
[x]   Fog value e5: COLOR=4494151a
[x]   Fog value e6: COLOR=44941519
[x]   Fog value e7: COLOR=44931518
[x]   Fog value e8: COLOR=44931417
[x]   Fog value e9: COLOR=44921416
[x]   Fog value ea: COLOR=44921415
[x]   Fog value eb: COLOR=44911414
[x]   Fog value ec: COLOR=44911413
[x]   Fog value ed: COLOR=44901412
[x]   Fog value ee: COLOR=44901411
[x]   Fog value ef: COLOR=448f1410
[x]   Fog value f0: COLOR=448f130f
[x]   Fog value f1: COLOR=448e130e
[x]   Fog value f2: COLOR=448e130d
[x]   Fog value f3: COLOR=448e130c
[x]   Fog value f4: COLOR=448d130b
[x]   Fog value f5: COLOR=448d130a
[x]   Fog value f6: COLOR=448c1309
[x]   Fog value f7: COLOR=448c1208
[x]   Fog value f8: COLOR=448b1207
[x]   Fog value f9: COLOR=448b1206
[x]   Fog value fa: COLOR=448a1205
[x]   Fog value fb: COLOR=448a1204
[x]   Fog value fc: COLOR=44891203
[x]   Fog value fd: COLOR=44891202
[x]   Fog value fe: COLOR=44881201
[x]   Fog value ff: COLOR=44881100
